client: correctly limit the number of outstanding requests

matklad opened the pull request:
Before, we used `TooManyOutstandingRequests` to signal this condition. The problem with that is that by the time we get to invoking user's callback, its already too late --- the code might have blown up with an assertion in `get_message`.

To fix this, check the number of outstanding requests in `get_message` instead. This also allows us to remove the error condition from the request callback altogether.

matklad committed:
client: correctly limit the number of outstanding requests

Before, we used `TooManyOutstandingRequests` to signal this condition.
The problem with that is that by the time we get to invoking user's
callback, its already too late --- the code might have blown up with an
assertion in `get_message`.

To fix this, check the number of outstanding requests in `get_message`
instead. This also allows us to remove the error condition from the
request callback altogether.

matklad commented on src/clients/c/tb_client/context.zig:
```diff
@@ -179,7 +178,6 @@ pub fn ContextType(
             );
             errdefer context.client.deinit(context.allocator);
 
-            context.messages_available = constants.client_request_queue_max;
```
I _think_ there was a subtle bug  here --- there's also `.register` request, which _won't_ be counted by this counter. New impl should be fine (but `.register` is an extra gotcha for the users to be aware of).

matklad commented on src/clients/c/tb_client/context.zig:
```diff
@@ -301,19 +297,14 @@ pub fn ContextType(
 
             // Signal to resume sending requests that was waiting for available messages.
-            if (self.messages_available == 1) self.thread.signal.notify();
+            if (self.client.messages_available == 1) self.thread.signal.notify();
```
This is subtle. In  the `Client.zig`, I actaully allow the users to retain messages beyond request-response itself. Eg, the user can do

```
const message = client.get_message();
some_global_var.message = message.ref();
client.request(message...)
```

So a request might become available either when a reply arrives, or when the user drops their outstanding reference. If the latter case happened here, that would have been a lost wakeup. 

However, tb_client fully controlls message and it doesn't retain references elsewher.

matklad commented on src/vsr/client.zig:
```diff
@@ -276,13 +282,23 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
+        ///
+        /// The caller must ensure that a message is available.
+        // FIXME: change back to get_message.
+        pub fn touch_get_message(self: *Self) *Message {
```
Renamed just to have the CI yell at me if I missed a callsite, will rename back before merge.

sentientwaffle commented on src/vsr/client.zig:
```diff
@@ -18,15 +20,11 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
                 user_data: u128,
                 operation: StateMachine.Operation,
-                results: Error![]const u8,
+                results: []const u8,
```
:tada:

sentientwaffle reviewed (commented):
LGTM once the temporary rename is fixed.

matklad commented:
Adjusted the node client as well. 

For all other clients, there's no other user-visible changes, because they are based on C client, which already:

- emits too-much-concurrency synchroneously
- transparentl retries failed requests

For node, the situation is different --- submitting request can now throw a new error, and request completion callback now always gets undefined as an error. I didn't change the signature of callbacks, though, because I think we are rewriting this to be based on the C client anyway?

matklad committed:
c_client: remove no longer necessary retry logic

We only retried on concurrency limit within Client, but that is now
synchronous.

matklad commented:
Ohhh, I can now actually kill the whole retry thing in our C client!

batiati reviewed (approved):
It was nice to get rid of the `retry` list, now the boundary between threads is much more simple to understand: application --> submitted at one side and submitted --> request at the other side.
